Support filter over indexes on nested fields

andrei-ionescu opened the pull request:
### What is the context for this pull request?

 - **Proposal**: #347
 - **Discussion**: #312
 - **Dependency**: #379 
 - Fixes: #347

### What changes were proposed in this pull request?

This PR adds support for filtering over nested fields using indexes created over nested fields. 

Given the `nestedDataset` dataset with schema

```
root
 |-- id: integer (nullable = true)
 |-- name: string (nullable = true)
 |-- nested: struct (nullable = true)
 |    |-- field1: string (nullable = true)
 |    |-- nst: struct (nullable = true)
 |    |    |-- field1: string (nullable = true)
 |    |    |-- field2: string (nullable = true)
```

and the following data

```
+---+-----+-----------------+
|id |name |nested           |
+---+-----+-----------------+
|2  |name2|[va2, [wa2, wb2]]|
|1  |name1|[va1, [wa1, wb1]]|
+---+-----+-----------------+
```

And the following search/filter query

```
df.filter(df("nested.nst.field1") === "wa1").select("id", "name", "nested.nst.field2")
```

The optimized and spark plans without index are

```
Project [id#100, name#101, nested#102.nst.field2]
+- Filter (isnotnull(nested#102) && (nested#102.nst.field1 = wa1))
   +- Relation[id#100,name#101,nested#102] parquet
```

```
Project [id#100, name#101, nested#102.nst.field2]
+- Filter (isnotnull(nested#102) && (nested#102.nst.field1 = wa1))
   +- FileScan parquet [id#100,name#101,nested#102] Batched: false, Format: Parquet, 
        Location: InMemoryFileIndex[file:/..../tableN2], PartitionFilters: [], 
        PushedFilters: [IsNotNull(nested)], ReadSchema:
        struct<id:int,name:string,nested:struct<field1:string,nst:struct<field1:string,field2:string>>>
```

The transformed optimized and spark plans should look like

```
Project [id#1, name#2, nested__nst__field2#3]
+- Filter (isnotnull(nested__nst__field1#0) && (nested__nst__field1#0 = wa1))
   +- Relation[nested__nst__field1#0,id#1,name#2,nested__nst__field2#3]
        Hyperspace(Type: CI, Name: idx_nested, LogVersion: 1)
```

```
Project [id#1, name#2, nested__nst__field2#3]
+- Filter (isnotnull(nested__nst__field1#0) && (nested__nst__field1#0 = wa1))
   +- FileScan parquet [nested__nst__field1#0,id#1,name#2,nested__nst__field2#3] Batched: false, 
        Format: Parquet, Location: InMemoryFileIndex[file:/..../spark_warehouse/indexes/idx_nested/v__=0],
        PartitionFilters: [], PushedFilters: [IsNotNull(nested__nst__field1)], ReadSchema: 
        struct<nested__nst__field1:string,id:int,name:string,nested__nst__field2:string>
```

## Complexities 

### Transforming the plan

Filters inside the plan must be modified to accomodate the index schema not the data schema - the flattened schema not the nested field. Instead of accessing the field with `GetStructField(GetStructField(AttributeReference))` it must directly access with `AttributeReference`.

Given the query plan

```
Project [id#100, name#101, nested#102.nst.field2]
+- Filter (isnotnull(nested#102) && (nested#102.nst.field1 = wa1))
   +- Relation[id#100,name#101,nested#102] parquet
```

The filter will be modified from

```
Filter (isnotnull(nested#102) && (nested#102.nst.field1 = wa1))
```

to

```
Filter (isnotnull(nested__nst__field1#0) && (nested__nst__field1#0 = wa1))
```

The projection from

```
Project [id#100, name#101, nested#102.nst.field2]
```

to

```
Project [id#1, name#2, nested__nst__field2#3]
```

The relation from 

```
Relation[id#100,name#101,nested#102] parquet
```

to

```
Relation[nested__nst__field1#0, id#1,name#2, nested__nst__field2#3] Hyperspace(Type: CI, 
  Name: idx_nested, LogVersion: 1)
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

- Existing unit and integration tests for not breaking the existing functionalities.
- Unit and integration test added for the new functionalities.

andrei-ionescu commented:
@sezruby, @imback82 Here is the second PR that adds support for filtering over nested fields with indexing.

sezruby commented on src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:
```diff
@@ -568,4 +589,56 @@ object RuleUtils {
+          extractSearchQuery(filter.condition, fieldName)
+        val fieldType = extractTypeFromExpression(exp, fieldName)
+        val attr = extractAttributeRef(exp, fieldName)
+        val exprId = getFieldPosition(index, escapedFieldName)
```
I think `exprId` should be unique. Could  you check `NamedExpression.newExprId` and utilize it if possible?

andrei-ionescu replied:
Yes. I'll look into it.

andrei-ionescu commented on src/main/scala/com/microsoft/hyperspace/index/rules/PlanUtils.scala:
```diff
@@ -0,0 +1,165 @@
+    val splits = name.split(".")
+    val elem = exp.flatMap {
+      case a: AttributeReference =>
+        if (splits.forall(s => a.name == s)) {
```
I don't know what you mean. Any string, even the empty string (`""`), when split it will return at least 1 item. And even for that one item we should iterate over all attributed and find it.

sezruby commented on src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:
```diff
@@ -568,4 +588,56 @@ object RuleUtils {
+      val escapedFieldName = SchemaUtils.prefixNestedFieldName(fieldName)
+      val attr = extractAttributeRef(exp, fieldName)
+      val fieldType = extractTypeFromExpression(exp, fieldName)
+      val exprId = getExprId(project, escapedFieldName)
```
Does this work as expected? I think it's because of 'transformDown', transformProject will be called before transforming the relation node. 
Could you add some test for filter index rule?

andrei-ionescu replied:
@sezruby: This functionality is already tested by existing UTs and E2Es. 

This is the correct approach: **Setting the `exprId` on the lowest leaf (aka: relation output) and when transforming each node get that expression id and use it when transforming that node**. So, the same expression id used in the relation is used in the tree otherwise there will be thrown analysis exception saying that the field (aka attribute) could not be found. I cannot set the `exprId` with a value on the relation, then in the filter and projection use other expression id (if I would use `newExprId` again). I have to take it from the relation that I previously created.

For example, given the `__hs_nested.nest.leaf.field`, in the relation I set it to a new number given by `newExprId` (ie: `#701`) and then on the filter to get another `exprId` (ie: `#801`), although at planning it will not crash it will throw exception at runtime.

This is how the correct plan should look like:

```
Project [__hs_nested.nested.nest.field#701]
+- Filter (isnotnull(__hs_nested.nested.nest.field#701) && (__hs_nested.nested.nest.field#701 = wa1))
   +- Relation[__hs_nested.nested.nest.field#701]
```

As you can see that `#701` is everywhere.

Now, regarding the `transformDown` and why it works, I guess the answer is that it is running multiple times.

I'll try switching to `transformUp` and see how it goes.

andrei-ionescu commented:
re-test

sezruby commented:
> Is there a mechanism to trigger the build once more on this PR? Something like commenting a keyword to trigger the build again?

You can trigger via "re-run" in `Checks` tab

andrei-ionescu commented:
@sezruby I don't see any "re-run" in the **Checks** tab. I don't have same right as you. 😃

andrei-ionescu commented:
@sezruby, @imback82 Can you have another look on the PR so that we can advance with it? Thanks!

sezruby commented:
Actually I'm waiting for #393 because of the conflict. 
@andrei-ionescu Could you review the change again? Since SchemaUtil is removed in that change, I wonder if it's ok or not.

andrei-ionescu commented:
@sezruby I didn't know that #393 is a prerequisite for this. If so, let's integrate feedback on that PR and get it merged so that I can advance with this one.

Remember we agreed to split the support for nested fields feature into multiple PRs so it can move faster. So, let's move faster 😁.

andrei-ionescu commented:
@sezruby, @imback82: I rebased the PR and resolved the conflicts and any issues. Just, FYI: `SchemaUtils` not needed in this PR. Please have another look ASAP.

andrei-ionescu commented:
@sezruby, @imback82: Ping? It's 2 days since my previous ping and last push. Please have a look on this PR.

andrei-ionescu commented:
@sezruby, @imback82: Can you re-run the build? It did fail due to:

```
sbt.InvalidComponent: Could not find required component 'xsbti'
```

Also, don't forget to review this PR. Thanks!

sezruby commented on src/main/scala/com/microsoft/hyperspace/index/rules/PlanUtils.scala:
```diff
@@ -0,0 +1,174 @@
+   * @param repl The replacement Spark Catalyst [[Expression]].
+   * @return A new Spark Catalyst [[Expression]].
+   */
+  def replaceInSearchQuery(
```
Could you revise the function name? seems it's irrelevant to search query

sezruby replied:
how about `replaceExpression`

sezruby commented on src/main/scala/com/microsoft/hyperspace/index/sources/FileBasedSourceProviderManager.scala:
```diff
@@ -109,6 +110,45 @@ class FileBasedSourceProviderManager(spark: SparkSession) {
+   * @param project Project to check if it's supported.
+   * @return True if the given project is a supported relation.
+   */
+  def isSupportedProject(project: Project, index: IndexLogEntry): Boolean = {
```
it's more like "hasNestedColumnWithProject" ?

andrei-ionescu replied:
I will find a better name.

sezruby commented on src/main/scala/com/microsoft/hyperspace/index/sources/FileBasedSourceProviderManager.scala:
```diff
@@ -109,6 +110,45 @@ class FileBasedSourceProviderManager(spark: SparkSession) {
+   * @param filter Filter to check if it's supported.
+   * @return True if the given project is a supported relation.
+   */
+  def isSupportedFilter(filter: Filter, index: IndexLogEntry): Boolean = {
```
ditto

sezruby commented on src/main/scala/com/microsoft/hyperspace/index/sources/FileBasedSourceProviderManager.scala:
```diff
@@ -109,6 +110,45 @@ class FileBasedSourceProviderManager(spark: SparkSession) {
+  def isSupportedProject(project: Project, index: IndexLogEntry): Boolean = {
+    val indexCols = index.indexedColumns ++ index.includedColumns
+    val resolvedIndexCols = indexCols.map(i => ResolverUtils.ResolvedColumn(i))
+    val hasNestedCols = resolvedIndexCols.exists(_.isNested)
```
How about return early if hasNestedCols is false?

sezruby commented on src/main/scala/com/microsoft/hyperspace/util/ResolverUtils.scala:
```diff
@@ -212,19 +161,106 @@ object ResolverUtils {
+   * @param name The column name resolved from an analyzed plan.
+   * @param isNested Flag to denote if this column is nested or not.
+   */
+  private[hyperspace] case class ResolvedColumn(name: String, isNested: Boolean) {
```
Could you move this function up to reduce diff?

andrei-ionescu replied:
I will but this is the way `scalafmt` optimises the code.

sezruby commented on src/main/scala/com/microsoft/hyperspace/util/ResolverUtils.scala:
```diff
@@ -212,19 +161,106 @@ object ResolverUtils {
+            case StructField(name, s: StructType, _, _) =>
+              name +: getColumnNameFromSchema(s, tail, resolver)
+            case StructField(_, _: ArrayType, _, _) =>
+              // TODO: Nested arrays will be supported later
```
Could you create an issue for this & link it? like other TODOs.

andrei-ionescu replied:
Support for Arrays of nested fields has the #372 proposal. I'll create a proposal for the maps too.

andrei-ionescu replied:
Created #411 for maps.

sezruby reviewed (commented):
Sorry for the delayed review. BTW could you add a simple test for the filter query in E2EHyperspaceRulesTest? Though it's tested in other PRs but need some minimum verifications for this change.

andrei-ionescu commented on src/main/scala/com/microsoft/hyperspace/index/rules/FilterIndexRule.scala:
```diff
@@ -111,20 +114,41 @@ object FilterIndexRule
+        (resolvedOutputColumnsOpt, resolvedFilterColumnsOpt) match {
+          case (Some(resolvedOutputColumns), Some(resolvedFilterColumns)) =>
+            val candidateIndexes = allIndexes.filter { index =>
+              indexCoversPlan(
```
The `resolve` methods used at lines 121 and 127 are different than the one inside `indexCoversPlan`. Inside `indexCoversPlan` it is not used the plan's function to resolve it, it only makes comparison between collections of fields.

andrei-ionescu commented:
@sezruby I integrated your feedback.

The build is still failing with the same previous error:

```
sbt.InvalidComponent: Could not find required component 'xsbti'
```

Can you check your build pipeline?

andrei-ionescu commented:
@sezruby, @imback82: Can you guys have **another look** and **fix the build** pipeline? It's been a while since this PR is up and I'm doing extra efforts to make the process as smooth as you need. Thanks!

sezruby commented:
@andrei-ionescu  I made an issue for build failure - https://github.com/actions/virtual-environments/issues/3146 

> I don't think we need another one. Or, is there a specific behaviour that you want to test?

I meant filter index test for nested column. You can change the column schema in an existing test or create a new test for it.

sezruby reviewed (commented):
I think not a just simple test, but some test set which is fully covered most of this PR is required as this is not simple enough to skip testing. 
@andrei-ionescu could you add some tests while build problem is fixing? Thanks

andrei-ionescu commented on src/main/scala/com/microsoft/hyperspace/index/sources/FileBasedSourceProviderManager.scala:
```diff
@@ -109,6 +110,49 @@ class FileBasedSourceProviderManager(spark: SparkSession) {
+   * @param project Project to check if it's supported.
+   * @return True if the given project is a supported relation.
+   */
+  def hasNestedColumns(project: Project, index: IndexLogEntry): Boolean = {
```
That was the best place I found. Do you have any suggestions?

sezruby replied:
Maybe RuleUtils .. ?

andrei-ionescu commented on src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala:
```diff
@@ -276,10 +280,29 @@ object RuleUtils {
+        val updatedOutput =
+          if (resolvedFields.isDefined && resolvedFields.get.exists(_.isNested)) {
+            indexFsRelation.schema.flatMap { s =>
+              relation.plan.output.find(a => s.name.contains(a.name)).map { a =>
```
Modified it to use the `ResolverUtils.ResolvedColumn` construct.

andrei-ionescu commented:
@sezruby, @imback82: I integrated your feedback and added an extra nested test in the `E2EHyperspaceRulesTest` called `E2E test for filter query on nested columns`.

Please have yet another look to this PR.

**Also, how can we speed things up?** This PR is more than a month old. I expected it to go faster after splitting the previous PR into multiple smaller ones.

sezruby commented:
@andrei-ionescu Sorry for the delay. but changes are not small or simple; need to spend some time to review it even with small PRs..

BTW... How about this?
original plan => normalized plan (nested filed => __hs_nested.xx) => trying to apply rule =>
1) if an index is applied => rename output
2) not applied => return original plan

In this way, maybe we don't need to change much of our existing code base or nested column related codes are not scattered.

andrei-ionescu commented:
@sezruby I can try to understand what you are saying. You want to say that in terms of impact the PR is not small - _it has a big impact PR_ - because in terms of number of lines modified in this PR are only 500+ modified lines. 

Even "_it has a big impact PR_" phrase is not true because the impact is minimised by the already present tests that guards the existing functionality of Hyperspace and there are lots of tests - a bit over 300 tests. They may not cover all the edge cases but it minimises the impact.

This second PR is taking more than a month to complete. This is way too much. 

Doing some math here... bringing the full support for nested fields in Hyperspace, with the current pace, it will take 6-12 months for it to land. IMHO **this process can be better**, I've seen it in other bigger projects with heavier features, with bigger PRs. 

@rapoth, @imback82: WDYT? What can be done to make this process faster? I did my part - the part that I could do - what else should I do to speed up the process?

andrei-ionescu commented:
@sezruby: Regarding this:

>original plan => normalized plan (nested filed => __hs_nested.xx) => trying to apply rule =>
>
>1. if an index is applied => denormalize (rename?) output
>2. not applied => return original plan
>
>In this way, maybe we don't need to change much of our existing code base or nested column related codes are not scattered.

If I understand your text correctly, we are doing exactly what you are saying. Given an index we renamed the output ([R275-R306](https://github.com/microsoft/hyperspace/pull/380/files#diff-8c312528287b51f613809d58a7b14b23c151d65794c099b653605795e0a738edR275-R306)), then we rename the field names in filters ([R310-R311](https://github.com/microsoft/hyperspace/pull/380/files#diff-8c312528287b51f613809d58a7b14b23c151d65794c099b653605795e0a738edR310-R311)) and projections ([R315-R316](https://github.com/microsoft/hyperspace/pull/380/files#diff-8c312528287b51f613809d58a7b14b23c151d65794c099b653605795e0a738edR315-R316)). 

I don't understand your request as for me it seems that the PR is doing exactly what you're requesting.

sezruby commented:
I meant out side of RuleUtils - normalize query plan from the beginning of apply function in FilterIndexRule & denormalize at the end of the apply function.

andrei-ionescu commented:
I don't think that is a good approach:
1. It needs to rename the fields in: output, filter and projections.
2. There is already quite some code manipulating the plans in `RuleUtils` so it seems to be the natural place to put it.
3. The same code would be used in hybrid scans and joins.
4. It would require passing through the whole plan twice each time (using `transformUp` and `transformDown` in both `FilterIndexRule` and `RuleUtils`) .
5. `JoinIndexRule` is different than `FilterIndexRule` and both use `RuleUtils` methods.

sezruby commented:
@andrei-ionescu 
1=> yep, but we could rollback(denormalize) nested columns after applying rules except for from index relation. 
2=> actually.. that's the problem. it's already complicated due to applying indexes & hybrid scan. We don't need to put everything into RuleUtils. I'll also have a plan to refactor hybrid scan because of the complexity.
3, 5=> if we flatten all the nested fields, I guess we don't need to change the code inside the logic of FilterIndexRule or JoinIndexRule or Hybrid Scan.  I currently work on this #403, so we will do flatten/unflatten once after that. (before then, I think it's okay normalize/denormalie for each FilterIndexRule & JoinIndexRule because ..
4=> I think that the benefit (code simplicity) outweighs the multiple iteration of plan. Spark also tries to apply twice (or multiple times) of each rule to confirm there's no transformation from the rule.

andrei-ionescu commented:
@sezruby 

> 1=> yep, but we could rollback(denormalize) nested columns after applying rules except for from index relation.

Sorry, but I don't understand what you want. Please provide examples with original plan and transformed plan.

> 2=> actually.. that's the problem. it's already complicated due to applying indexes & hybrid scan. We don't need to put everything into RuleUtils. I'll also have a plan to refactor hybrid scan because of the complexity.

I can understand this and I'm ok with splitting the `RuleUtils` big file into smaller ones.

> 3, 5=> if we flatten all the nested fields, I guess we don't need to change the code inside the logic of FilterIndexRule or JoinIndexRule or Hybrid Scan. I currently work on this #403, so we will do flatten/unflatten once after that. (before then, I think it's okay normalize/denormalie for each FilterIndexRule & JoinIndexRule because ..

I don't understand the flatten process you are talking - for me is not clear if you are referring to getting from `nested.field` into `__hs_nested.nested.field` or is other process.

I've seen the your "HyperspaceOneRule" proposal and PR and it looks great. I've seen how it's being split at this moment and that the `RuleUtils` file remains the same. 

But, it is **_NOT FAIR_** to postpone and stall this PR for that one, neither making me address all the code conflicts - although I'm willing to address all of them if your PR gets merged fast (fast == next day).  

> 4=> I think that the benefit (code simplicity) outweighs multiple iterations of plan. Spark also tries to apply twice (or multiple times) of each rule to confirm there's no transformation from the rule.

Sorry, I cannot see the "code simplicity" you are talking about. This may be because I don't understand what you want as I already said.

sezruby commented:
@andrei-ionescu Ok let me explain more about my suggestion:

queryPlan = original  query plan before transformation

1) nomalizedPlan = flatten + prefix all nested columns in queryPlan
```
       *   Project [id#100, name#101, nested#102.nst.field2]
       *   +- Filter (isnotnull(nested#102) && (nested#102.nst.field1 = wa1))
       *      +- Relation[id#100,name#101,nested#102] parquet
```
to
```
       *   Project [id#100, name#101, __hs_nested.nested.nst.field2#3]
       *   +- Filter (isnotnull(__hs_nested.nested.nst.field1#0) && (__hs_nested.nested.nst.field1#0 = wa1))
       *      +- Relation[__hs_nested.nested.nst.field1#0,id#100,name#101, __hs_nested.nested.nst.field2#3] parquet
```

2) try to apply rules to normalizedPlan, so
```
  override def apply(plan: LogicalPlan): LogicalPlan = {
    ...
    plan transformDown {
```
will be:
```
  override def apply(plan: LogicalPlan): LogicalPlan = {
    val normalizedPlan = normalizeNestedColumns(plan)
    val transformedPlan = normalizedPlan transformDown {
   ..
```

3) after applying rules, 

```
// if there's no transformation, return original queryPlan
if (transformedPlan.equals(normalizedPlan)) return plan

// Otherwise, 
denormalizeNestedColumns(transformedPlan)
```

In denormalizedNestedColumns, we could check if each normalized column is from index relation or not and replace it with the original query plan tree if it's not from the index relation.

===
BTW should we rename the output column or no?
For example:
```
Project [id#1, name#2, __hs_nested.nested.nst.field2#3]
...
``
What is the result column name of " __hs_nested.nested.nst.field2#3"?

andrei-ionescu commented:
@sezruby 
> BTW should we rename the output column or no?

Yes. We should. I have that in a later PR and forgot about it.

andrei-ionescu commented:
@sezruby 

> 1. nomalizedPlan = flatten + prefix all nested columns in queryPlan

This requires to go through the whole plan with `transformDown` or `transformUp` and: a) change the relation (schema & attribute refs name and exprId)  b) change the filter (attribute refs name and exprId that needs to match the relation) c) change the project (attribute refs name and exprId that also needs to match the relation one). This is how I understand to get the "normalizedPlan" as you call it.

> 2. try to apply rules to normalizedPlan, so

Given the transformation above, what else is left to apply? I don't understand this "apply rules to normalised plan". Which rules are you referring to? For me a rule is a matching of a plan node that will get transformed.

If `normalizeNestedColumns(plan)` does run a `transformDown` and changes the plan altogether, doing `val transformedPlan = normalizedPlan transformDown` doesn't make any sense. 

> 3. after applying rules,
>
> ```
> if (transformedPlan.equals(normalizedPlan)) return plan
> ```

This does not make any sense either. You should check against the original `plan` not `transformed` vs `normalizedPlan` and return `plan`.

I guess I don't understand what "apply rules" does. The current Hyperspace approach is to FIRST select a suitable index over all the plan and after that is found start the plan transformation based in the columns inside the index. Is your PR - "HyperspaceOneRule" - changing this process?

sezruby commented:
Sorry about rule"s".. I meant it's FilterIndexRule. I'd like to separate "applying index" and "handling nested columns".

normalizedNestedColumns only transforms nested columns - without considering indexes. 
(though not sure this is possible - creating relation node using different column schema during the transformation)

andrei-ionescu commented:
@sezruby

> separate "applying index" and "handling nested columns"

What's the use case for this separation? Will there be ever a case where we will do `applyIndex(plan)` and then not iterate over the plan changing attributes where needed?
  
In the context of nested fields `applyIndex(plan)` without transforming the rest of the plan makes no sense and will NOT work.

> normalizedNestedColumns only transforms nested columns - without considering indexes.

IMHO I don't think this is a good approach - to transform the all the nested columns without considering what's in the index. I think the opposite: we should change only what matches for changes, not everything.

From your comments, reading between the lines, I understand that you are NOT happy that supporting nested fields adds so many changes in the current `RuleUtils` file in the `transformPlanToUseIndexOnlyScan` and `transformPlanToUseHybridScan` methods. If this is the case I propose to make the `RuleUtils` a case class and name it `BaseRuleHelper`. Then out of it, create the `NestedFieldsRuleHelper` that will extend it and override the needed methods. `FilterIndexRule` and `JoinIndexRule` should detect the case and when there are nested fields involved, it should use the `NestedFieldsRuleHelper` otherwise fall back to `BaseRuleHelper`.

@sezruby, @imback82 WDYT?

sezruby commented:
@andrei-ionescu That sounds a good alternative. It's already complicated with source provider & hybrid scan (need to decrypt to read it). Let's modulize it as much as possible.. 
Sorry for the late suggestion, but it came up after reading the code over and over while reviewing.

andrei-ionescu commented:
@imback82 Are you ok with it? I'm just making sure that I'm not starting working on something that is not agreed by you too. Otherwise my effort may be just wasted time.

sezruby commented:
@clee704 could you review this PR & above few discussion comments & share some opinion on it?

andrei-ionescu commented:
@clee704, @imback82, @sezruby Can I have an answer on my question and proposal above, so that I can advance with this PR?

sezruby commented:
@andrei-ionescu I believe that it would be better than the current approach - if we need to support other plan type, it will be getting bigger & complicated (e.g. nested array, map, ..) and I'm the main reviewer before @imback82, so feel free to proceed it. 
@clee704 is relatively new to Hyperspace so it might need some time to review this change & @imback82 is so busy this week..

andrei-ionescu commented:
@sezruby Thanks for green light. I understand that your approval should be enough for this PR, once that I change the implementation. I don't want to be in the position that someone else disagrees with it and I have to start over one more. You know: once bitten, twice shy.

rapoth commented:
@andrei-ionescu Thanks for double-checking! Please go ahead with @sezruby's recommendation - I agree that's a sound proposal and it will help us all in the long run. :)

andrei-ionescu commented:
@rapoth, @sezruby Thanks for agreement. I'll start grinding on it ASAP.

andrei-ionescu commented on src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala:
```diff
@@ -60,17 +60,16 @@ object JoinIndexRule
         getBestIndexPair(l, r, condition)
           .map {
             case (lIndex, rIndex) =>
+              val ruleHelper = new BaseRuleHelper(spark)
```
Changes in this file are mainl due to `RuleUtils` changes.

andrei-ionescu commented on src/test/scala/com/microsoft/hyperspace/index/rules/RuleUtilsTest.scala:
```diff
@@ -90,16 +90,17 @@ class RuleUtilsTest extends HyperspaceRuleSuite with SQLHelper {
-    val t2Relation = RuleUtils.getRelation(spark, t2ProjectNode).get
-    assert(RuleUtils.getCandidateIndexes(spark, allIndexes, t1Relation).length === 3)
-    assert(RuleUtils.getCandidateIndexes(spark, allIndexes, t2Relation).length === 2)
+    val ruleHelper = new BaseRuleHelper(spark)
```
Changes in here are due to the fact the `RuleUtils` has been renamed and transformed into a class.

andrei-ionescu commented:
@sezruby: Here is the refactor:

1. Renamed `RuleUtils` into `BaseRuleHelper` and made it a class
2. Created `NestedRuleHelper` that extends `BaseRuleHelper`
3. Added "detection code" into `FilterIndexRule`
4. Kept the new nested field E2E test
5. Added index nested field rename back to original name

Please have a look.

sezruby commented on src/main/scala/com/microsoft/hyperspace/index/rules/NestedRuleHelper.scala:
```diff
@@ -0,0 +1,501 @@
+          new ParquetFileFormat,
+          Map(IndexConstants.INDEX_RELATION_IDENTIFIER))(spark, index)
+
+        val resolvedFields = ResolverUtils.resolve(
```
Could you define some functions (outputFields, transformProject..) and override them rather than overriding whole transformPlanToUse*?
In this way, it's hard to maintain changes in transformPlanToUse*

sezruby commented on src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala:
```diff
@@ -108,11 +107,12 @@ object JoinIndexRule
 
-    lazy val optLeftRel = RuleUtils.getRelation(spark, l)
-    lazy val optRightRel = RuleUtils.getRelation(spark, r)
+    lazy val ruleHelper = new BaseRuleHelper(spark)
```
Could you let some of utility functions like getRelation in `object RuleUtils`? to avoid unnecessary instance creation

andrei-ionescu replied:
The only method I created as outside of the class instantiation is `BaseRuleHelper.isIndexApplied` and is part of the companion object. 

All other functions do use `spark: SparkSession` inside their logic and  because of it they are now used through instantiation. 

Regarding `getRelation` that also uses `spark` inside it but we can extract that part and provide a `HyperspaceContext` parameter.

clee704 commented on src/main/scala/com/microsoft/hyperspace/index/rules/FilterIndexRule.scala:
```diff
@@ -53,19 +54,24 @@ object FilterIndexRule
-                  originalPlan,
-                  useBucketSpec = HyperspaceConf.useBucketSpecForFilterRule(spark),
-                  useBucketUnionForAppended = false)
+              // Detect which whether the index contains nested fields.
```
```suggestion
              // Detect whether the index contains nested fields.
```

clee704 commented on src/main/scala/com/microsoft/hyperspace/index/rules/NestedRuleHelper.scala:
```diff
@@ -0,0 +1,262 @@
+    if (hasNestedCols) {
+      val filterFields = extractNamesFromExpression(filter.condition).toKeep.toSeq
+      val resolvedFilterFields = filterFields.map(ResolverUtils.ResolvedColumn(_))
+      resolvedFilterFields.exists(i => indexCols.exists(j => j == i || j.name == i.name))
```
Is it possible that j.name != i.name but j == i?

clee704 commented on src/main/scala/com/microsoft/hyperspace/index/rules/NestedRuleHelper.scala:
```diff
@@ -0,0 +1,262 @@
+   * @param project The project that needs to be transformed.
+   * @return The transformed project with support for nested indexed fields.
+   */
+  protected[rules] def transformProject(project: Project): Project = {
```
Does it work for nested fields in expressions? Example: `select my.nested.field1 + my.nested.field2 from mytab`

clee704 replied:
Also, it seems to be a good candidate for a unit test

clee704 commented on src/main/scala/com/microsoft/hyperspace/index/rules/NestedRuleHelper.scala:
```diff
@@ -0,0 +1,262 @@
+   * @param filter The filter that needs to be transformed.
+   * @return The transformed filter with support for nested indexed fields.
+   */
+  protected[rules] def transformFilter(filter: Filter): Filter = {
```
seems to be a good candidate for a unit test

clee704 commented on src/main/scala/com/microsoft/hyperspace/index/rules/NestedRuleHelper.scala:
```diff
@@ -0,0 +1,262 @@
+   * @param project Project to check if it's supported.
+   * @return True if the given project is a supported relation.
+   */
+  protected[rules] def hasNestedColumns(project: Project, index: IndexLogEntry): Boolean = {
```
seems to be a good candidate for a unit test

clee704 commented on src/main/scala/com/microsoft/hyperspace/index/rules/NestedRuleHelper.scala:
```diff
@@ -0,0 +1,262 @@
+   * @param filter Filter to check if it's supported.
+   * @return True if the given project is a supported relation.
+   */
+  protected[rules] def hasNestedColumns(filter: Filter, index: IndexLogEntry): Boolean = {
```
seems to be a good candidate for a unit test

clee704 commented on src/main/scala/com/microsoft/hyperspace/index/rules/NestedRuleHelper.scala:
```diff
@@ -0,0 +1,262 @@
+   * @param fieldName The name of the field to search for.
+   * @return An [[ExprId]] if that could be found in the plan otherwise [[None]].
+   */
+  private def getExprId(plan: LogicalPlan, fieldName: String): Option[ExprId] = {
```
seems to be a good candidate for a unit test

clee704 commented on src/main/scala/com/microsoft/hyperspace/index/rules/PlanUtils.scala:
```diff
@@ -0,0 +1,198 @@
+   * @param exp The Spark Catalyst expression from which to extract names.
+   * @return A set of distinct field names.
+   */
+  def extractNamesFromExpression(exp: Expression): ExtractedNames = {
```
seems to be a good candidate for a unit test

andrei-ionescu replied:
I will add tests to all methods in here.

clee704 commented on src/main/scala/com/microsoft/hyperspace/index/rules/PlanUtils.scala:
```diff
@@ -0,0 +1,198 @@
+
+import com.microsoft.hyperspace.util.ResolverUtils
+
+object PlanUtils {
```
The name sounds general, but the actual code seems mostly about nested fields?

andrei-ionescu replied:
The code work for any kind of field. Is needed by the nested fields feature because it needs to do more transformations.

clee704 commented on src/main/scala/com/microsoft/hyperspace/index/rules/PlanUtils.scala:
```diff
@@ -0,0 +1,198 @@
+      }
+    }
+
+    var toRemove = Seq.empty[String]
```
This is to remove the unnecessary isnull check, right? If someone without the knowledge that this code is about supporting nested fields, the meaning of this name could be vague.

andrei-ionescu replied:
No. Given this plan:
```
Project [Date#89, nested#94.leaf.cnt AS cnt#336, nested#94.leaf.id AS id#337]
+- Filter ((isnotnull(nested#94) && (nested#94.leaf.cnt > 10)) && (nested#94.leaf.id = leaf_id9))
   +- Relation[Date#89,RGUID#90,Query#91,imprs#92,clicks#93,nested#94] parquet
```
It will be transformed into:
```
Project [Date#380 AS Date#381, __hs_nested.nested.leaf.cnt#379 AS cnt#382, __hs_nested.nested.leaf.id#378 AS id#383]
+- Filter ((__hs_nested.nested.leaf.cnt#379 > 10) && (__hs_nested.nested.leaf.id#378 = leaf_id9))
   +- Relation[__hs_nested.nested.leaf.id#378,__hs_nested.nested.leaf.cnt#379,Date#380] Hyperspace(Type: CI, Name: filterNestedIndex, LogVersion: 1)
```
The `isnotnull(nested#94)` is no longer suitable and is removed in the `transformFilter` method because:
- The `nested` field name is not part of the index (index contains `__hs_nested.nested.leaf.cnt` and `__hs_nested.nested.leaf.id`)
- The `isnotnull` construct checks for a nested field to not be null because trying to access the leaves when it's null would end up in exceptions
- The values stored in the index are not nested, they are flat
- In the example above, we cannot just rename the `isnotnull` attribute reference as we have two fields instead of one (`__hs_nested.nested.leaf.cnt` and `__hs_nested.nested.leaf.id`)

sezruby replied:
Can we add both isnotnull(__hs_nested.nested.leaf.cnt) & isnotnull(__hs_nested.nested.leaf.id)? 
Since it seems Spark automatically adds `isnotnull` for project columns and filter condition columns.

andrei-ionescu replied:
I think we can. I'll try it.

clee704 commented:
Maybe a dumb question, but why should we rename the fields? Can't we just keep the original nested structures in index data?

sezruby commented on src/main/scala/com/microsoft/hyperspace/index/rules/PlanUtils.scala:
```diff
@@ -0,0 +1,198 @@
+  def replaceInSearchQuery(
+      parent: Expression,
+      needle: Expression,
+      repl: Expression): Expression = {
```
`replacement` for more clarity?

sezruby commented:
> Maybe a dumb question, but why should we rename the fields? Can't we just keep the original nested structures in index data?

@clee704 That's a good question. One reason could be that we store index data as "indexed columns" ++ "included columns". For example, if indexed columns are `nested.a, b` and included columns are `nested.c, d`, then it's not possible to define the column schema with that order because of the same field name - `nested`. (But I wonder the order of columns in index data does really matter. )
And maybe another reason might be there're several issues on handling nested columns itself in Spark & to minimize design change in Hyperspace.

andrei-ionescu commented:
@sezruby, @clee704: The datasets I use to create indexes over contains deep nested fields with many children.

For example this is a half of the schema of one of my datasets:

```
root
 |-- ts: timestamp (nullable = true)
 |-- id: string (nullable = true)
 |-- xs: struct (nullable = true)
 |    |-- an: struct (nullable = true)
 |    |    |-- ev1_100: struct (nullable = true)
 |    |    |    |-- e1: struct (nullable = true)
 |    |    |    |    |-- value: double (nullable = true)
 |    |    |    |-- e2: struct (nullable = true)
 |    |    |    |    |-- value: double (nullable = true)
 |    |    |      ...
 |    |    |    |-- e100: struct (nullable = true)
 |    |    |    |    |-- value: double (nullable = true)
 |    |    |-- ev101_200: struct (nullable = true)
 |    |    |    |-- e101: struct (nullable = true)
 |    |    |    |    |-- value: double (nullable = true)
 |    |    |      ...
 |    |    |    |-- e200: struct (nullable = true)
 |    |    |    |    |-- value: double (nullable = true)
 |    |    |-- ev201_300: struct (nullable = true)
 |    |    |    |-- e201: struct (nullable = true)
 |    |    |    |    |-- value: double (nullable = true)
 |    |    |      ...
 |    |    |    |-- e300: struct (nullable = true)
 |    |    |    |    |-- value: double (nullable = true)
 |    |    |-- ev301_400: struct (nullable = true)
 |    |    |    |-- e301: struct (nullable = true)
 |    |    |    |    |-- value: double (nullable = true)
 |    |    |      ...
 |    |    |    |-- e400: struct (nullable = true)
 |    |    |    |    |-- value: double (nullable = true)
 |    |    |-- dim: struct (nullable = true)
 |    |    |    |-- v: struct (nullable = true)
 |    |    |    |    |-- v1: string (nullable = true)
 |    |    |    |    |-- v2: string (nullable = true)
 |    |    |    |      ...
 |    |    |    |    |-- v99: string (nullable = true)
 |    |    |    |    |-- v100: string (nullable = true)
 |    |    |    |-- p: struct (nullable = true)
 |    |    |    |    |-- p1: string (nullable = true)
 |    |    |    |    |-- p2: string (nullable = true)
 |    |    |    |      ...
 |    |    |    |    |-- p_99: string (nullable = true)
 |    |    |    |    |-- p_100: string (nullable = true)
 |    |    |    |-- hiers: struct (nullable = true)
 |    |    |    |    |-- h1: struct (nullable = true)
 |    |    |    |    |    |-- values: array (nullable = true)
 |    |    |    |    |    |    |-- element: string (containsNull = true)
 |    |    |    |    |    |-- delimiter: string (nullable = true)
 |    |    |    |      ...
 |    |    |    |-- lists: struct (nullable = true)
 |    |    |    |    |-- l1: struct (nullable = true)
 |    |    |    |    |    |-- list: array (nullable = true)
 |    |    |    |    |    |    |-- element: struct (containsNull = true)
 |    |    |    |    |    |    |    |-- value: string (nullable = true)
 |    |    |    |    |-- l2: struct (nullable = true)
 |    |    |    |    |    |-- list: array (nullable = true)
 |    |    |    |    |    |    |-- element: struct (containsNull = true)
 |    |    |    |    |    |    |    |-- value: string (nullable = true)
 |    |    |    |      ...
 |-- sch: struct (nullable = true)
 |    |-- se: string (nullable = true)
 |    |-- keys: string (nullable = true)
```

I want to index over the `xs.an.dim.p.p_99` nested field. Are you proposing to keep the whole `xs` struct in the index?

If so, I don't think that is efficient at all and I'll give a few reasons:
1. **Storage cost** — It will take a lot of space
2. **Performance cost** — It will be less usable due to response time when used in queries
3. **Management cost** — It will be harder to maintain up to date

Performance and cost wise I can say that the current approach — _extracting/flattening the nested field_ — is what is needed in Hyperspace even though the implementation code gets more complex.

FYI: I'm currently a bit busy with some other work assignment but I'll get back to integrating your feedback soon.

clee704 commented:
@andrei-ionescu Since we store data in a parquet format, which is efficient in storing nested data in a columnar format, shouldn't it matter? For a parquet file, a dataset with a single column "a" or a single nested column "a.b.c.d.e" should make no difference, I guess.

andrei-ionescu commented:
It does matter! Even if parquet stores data in an efficient columnar way, there are issues with parquet and nested fields in the way the predicate push down works - or to say it better, it doesn't work. This is the **performance cost**. There is the **storage cost** too that is not negligible either - if you are forced to store half of the dataset in an index. Lastly but not the least, the **management cost** (rebuild index, incremental refresh, etc) will incur additional time that will make index harder update and keep in sync with the data itself - if the index is not easily (fast & efficient) updatable then it's not usable in these _big data fast moving datasets_. These costs are not acceptable for the cases of big nested datasets.

@rapoth, @imback82, @sezruby, @apoorvedave1, @clee704: Here is an article I wrote on what Adobe is doing on big data and what's the scale we work with: https://medium.com/adobetech/high-throughput-ingestion-with-iceberg-ccf7877a413f. Or if you prefer video here is a presentation on the same matter: https://www.dremio.com/subsurface/high-frequency-small-files-vs-slow-moving-datasets.

imback82 commented:
I don't think we can "bucket by" nested fields either. BTW, predicate pushdown for nested fields should be supported in Spark 3.0: https://github.com/apache/spark/commit/cb0db213736de5c5c02b09a2d5c3e17254708ce1

@clee704 is your concerned addressed?

andrei-ionescu commented:
That's right @imback82! I totally forgot about the bucketing done in Hyperspace. That is not possible over nested fields.

clee704 commented:
> I want to index over the xs.an.dim.p.p_99 nested field. Are you proposing to keep the whole xs struct in the index?

Can't we just store the indexed/included nested fields, ignoring others?

andrei-ionescu commented:
@sezruby, @clee704, @imback82 I'll get back to this PR very soon. I've been side tracked by another assignment but I'll get back to it integrating the feedback and adding the needed UTs.

andrei-ionescu commented:
@imback82, @sezruby, @clee704: Sorry for being out of this PR for a few weeks. Now, I got some time and I'm back to it. Here are the changes:

- Rebased it over Spark 3 support and resolved conflicts
- Integrated your feedback
- Added `PlanUtilsTest` and `NestedRuleHelperTest` tests to validate changes on the plan

Please have another look.

sezruby commented on src/main/scala/com/microsoft/hyperspace/index/rules/BaseRuleHelper.scala:
```diff
@@ -0,0 +1,574 @@
+   * @param relation Relation with which given indexes are compared.
+   * @return Active indexes built for this plan.
+   */
+  def getCandidateIndexes(
```
Please address this comment - to reduce the diff

andrei-ionescu replied:
@sezruby 

1) I rather not put back `getCandidateIndexes` as the method requires spark being given to it. The current implementation is consistent as the methods that don't require spark are in `RuleUtils` while the other are part of the helper instances.
2) There 3000 lined due to extensive tests that I added to the verify the plan transformation. In a way or another someone should review it.
3) "_Frankly speaking_", I already did a lot of extra work complying with your requests of splitting in multiple PRs. Now, once more this request to split PR to this same feature. This is very frustrating for me and looks like you using any opportunity to push this back. I would understand the split into multiple PRs for new features but I though we did come to an agreement at the time we decided to split the feature in 4-5 other PRs.

sezruby replied:
1=>
It will be  refactored out after rule refactoring. Just put it back to reduce the diff...
You can use spark from FilterIndexRule / JoinIndexRule, as before.
Also no need to `getCandidate` function in this class. (IndexPlanApplyHelper ?) - they have different role.

2/3=>
I understand your frustration, but the changes modify wide range of codebase and also core part.
This refactoring was not considered at the time we agreed and that's why this change became huge.
And you know small & explicit changes can be merged quickly, in general.. and splitting PR is also helping to trace the history. 

Sorry but I think we should do that. If you are not up for it, I'll push a refactoring PR(no nested related code) based on your PR, until early next week.

andrei-ionescu replied:
@sezruby I want to correct you. The code does not add lots of changes. If you remember - and to help you remember, here is a previous discussion a month back, https://github.com/microsoft/hyperspace/pull/380#issuecomment-820231714 - we agreed to have separate classes for the already existing code and the nested features. So now we have `BaseRuleHelper` and `NestedRuleHelper`. From the old `RuleUtils` I extracted all the methods that need `spark` into the `BaseRuleHelper` and kept the methods that don't need `spark` in `RuleUtils` as static methods. So, from the old `RuleUtils` now we have 2 files: a stripped down `RuleUtils` and the `BaseRuleHelper`. The classes have the code unchanged from the old `RuleUtils`, only the method signatures are changed now that `spark` is a property inside the instance.

The new things are inside `NestedRuleHelper` which extends some methods from `BaseRuleHelper` to accommodate the nested fields support.

This is exactly what we agreed on that comment.

There is no other refactor here than this what I explained above refactor that does NOT make any sense without the next changes.

I propose to merge try merge this and you create the changes that you want on top of it.

@imback82, @rapoth: What is your opinion?

sezruby commented on src/main/scala/com/microsoft/hyperspace/index/rules/FilterIndexRule.scala:
```diff
@@ -53,19 +54,24 @@ object FilterIndexRule
+              // Detect whether the index contains nested fields.
+              val indexHasNestedColumns = (index.indexedColumns ++ index.includedColumns)
+                .exists(ResolverUtils.ResolvedColumn(_).isNested)
+              val ruleHelper = if (indexHasNestedColumns) {
```
instead of this, how about defining 

```scala
object IndexRelationApplyHelper {
  def apply(index: IndexLogEntry): IndexRelationApplyHelper
}
```

andrei-ionescu commented:
@imback82, @rapoth: How can we get this moving? It's more than 3 months now...
